### ISSUE TITLE
Heavy I/O causes load in excess of 50, up to 100. 

### ISSUE THREAD

**Issue by asbach** [zfsonlinux/zfs#3061]
I'm running an Ubuntu 14.04 host as a Samba/NFS server using 8 drives in a RAID10 config. Since the last zfs update (0.6.3-5~trusty) I get loads up to 100 during heavy I/O when for example tivoli backups are running. arc_adapt and kswapd0 are causing heavy cpu load and perf shows most of my cycles going to _raw_spin_lock and kmem_cache_shrink. I've tried setting my spl_kmem_cache_slab_lmit to 16384. This has helped a little with the problem, but I'm still experiencing these heavy loads. Before the zfs update, load would hardly ever be above 2.0.

```
zpool status
  pool: poolTank
 state: ONLINE
  scan: scrub repaired 0 in 38h26m with 0 errors on Mon Jan 26 09:56:58 2015
config:

        NAME                        STATE     READ WRITE CKSUM
        poolTank                    ONLINE       0     0     0
          mirror-0                  ONLINE       0     0     0
            scsi-35000cca03b3c9b74  ONLINE       0     0     0
            scsi-35000cca03b49ddcc  ONLINE       0     0     0
          mirror-1                  ONLINE       0     0     0
            scsi-35000cca03b49de14  ONLINE       0     0     0
            scsi-35000cca03b49dfd4  ONLINE       0     0     0
          mirror-2                  ONLINE       0     0     0
            scsi-35000cca03b4e1748  ONLINE       0     0     0
            scsi-35000cca03b4e3ce4  ONLINE       0     0     0
          mirror-3                  ONLINE       0     0     0
            scsi-35000cca03b4e400c  ONLINE       0     0     0
            scsi-35000cca03b4e8278  ONLINE       0     0     0

errors: No known data errors
```

```
CPUs:           2
Memory:         32GB
VM/Hypervisor:        no
ECC mem:        yes
Distribution:       Ubuntu 14.04
Kernel version:     Linux psi 3.13.0-44-generic #73-Ubuntu SMP Tue Dec 16 00:22:43 UTC 2014 x86_64 x86_64 x86_64 GNU/Linux
SPL/ZFS source:     Packages from ubuntu trusty stable ppa
SPL/ZFS version:    dmesg:[   19.612727] SPL: Loaded module v0.6.3-3~trusty
dmesg:[   19.814593] ZFS: Loaded module v0.6.3-5~trusty, ZFS pool version 5000, ZFS filesystem version 5 filesystem version 5       
System services:    Samba/NFS/Kerberos/LDAP
Short  description: Heavy IO causes loads in excess of 50, reaching up to 100.
```

module parameters:

```
spl_debug_mask                                    188
spl_debug_mb                                      -1
spl_debug_panic_on_bug                            0
spl_debug_printk                                  60
spl_debug_subsys                                  18446744073709551615
spl_hostid                                        1015427281
spl_hostid_path                                   /etc/hostid
spl_kmem_cache_expire                             2
spl_kmem_cache_kmem_limit                         1024
spl_kmem_cache_max_size                           32
spl_kmem_cache_obj_per_slab                       16
spl_kmem_cache_obj_per_slab_min                   8
spl_kmem_cache_reclaim                            0
spl_kmem_cache_slab_limit                         16384
spl_taskq_thread_bind                             0
l2arc_feed_again                                  1
l2arc_feed_min_ms                                 200
l2arc_feed_secs                                   1
l2arc_headroom                                    2
l2arc_headroom_boost                              200
l2arc_nocompress                                  0
l2arc_noprefetch                                  1
l2arc_norw                                        0
l2arc_write_boost                                 8388608
l2arc_write_max                                   8388608
metaslab_debug_load                               0
metaslab_debug_unload                             0
spa_asize_inflation                               24
spa_config_path                                   /etc/zfs/zpool.cache
zfetch_array_rd_sz                                1048576
zfetch_block_cap                                  256
zfetch_max_streams                                8
zfetch_min_sec_reap                               2
zfs_arc_grow_retry                                5
zfs_arc_max                                       10737418240
zfs_arc_memory_throttle_disable                   1
zfs_arc_meta_limit                                0
zfs_arc_meta_prune                                1048576
zfs_arc_min                                       0
zfs_arc_min_prefetch_lifespan                     250
zfs_arc_p_aggressive_disable                      1
zfs_arc_p_dampener_disable                        1
zfs_arc_shrink_shift                              5
zfs_autoimport_disable                            1
zfs_dbuf_state_index                              0
zfs_deadman_enabled                               1
zfs_deadman_synctime_ms                           1000000
zfs_dedup_prefetch                                1
zfs_delay_min_dirty_percent                       60
zfs_delay_scale                                   500000
zfs_dirty_data_max                                3373851443
zfs_dirty_data_max_max                            8434628608
zfs_dirty_data_max_max_percent                    25
zfs_dirty_data_max_percent                        10
zfs_dirty_data_sync                               67108864
zfs_disable_dup_eviction                          0
zfs_expire_snapshot                               300
zfs_flags                                         1
zfs_free_min_time_ms                              1000
zfs_immediate_write_sz                            32768
zfs_mdcomp_disable                                0
zfs_nocacheflush                                  0
zfs_nopwrite_enabled                              1
zfs_no_scrub_io                                   0
zfs_no_scrub_prefetch                             0
zfs_pd_blks_max                                   100
zfs_prefetch_disable                              0
zfs_read_chunk_size                               1048576
zfs_read_history                                  0
zfs_read_history_hits                             0
zfs_recover                                       0
zfs_resilver_delay                                2
zfs_resilver_min_time_ms                          3000
zfs_scan_idle                                     50
zfs_scan_min_time_ms                              1000
zfs_scrub_delay                                   4
zfs_send_corrupt_data                             0
zfs_sync_pass_deferred_free                       2
zfs_sync_pass_dont_compress                       5
zfs_sync_pass_rewrite                             2
zfs_top_maxinflight                               32
zfs_txg_history                                   0
zfs_txg_timeout                                   5
zfs_vdev_aggregation_limit                        131072
zfs_vdev_async_read_max_active                    3
zfs_vdev_async_read_min_active                    1
zfs_vdev_async_write_active_max_dirty_percent     60
zfs_vdev_async_write_active_min_dirty_percent     30
zfs_vdev_async_write_max_active                   10
zfs_vdev_async_write_min_active                   1
zfs_vdev_cache_bshift                             16
zfs_vdev_cache_max                                16384
zfs_vdev_cache_size                               0
zfs_vdev_max_active                               1000
zfs_vdev_mirror_switch_us                         10000
zfs_vdev_read_gap_limit                           32768
zfs_vdev_scheduler                                noop
zfs_vdev_scrub_max_active                         2
zfs_vdev_scrub_min_active                         1
zfs_vdev_sync_read_max_active                     10
zfs_vdev_sync_read_min_active                     10
zfs_vdev_sync_write_max_active                    10
zfs_vdev_sync_write_min_active                    10
zfs_vdev_write_gap_limit                          4096
zfs_zevent_cols                                   80
zfs_zevent_console                                0
zfs_zevent_len_max                                128
zil_replay_disable                                0
zil_slog_limit                                    1048576
zio_bulk_flags                                    0
zio_delay_max                                     30000
zio_injection_enabled                             0
zio_requeue_io_start_cut_in_line                  1
zvol_inhibit_dev                                  0
zvol_major                                        230
zvol_max_discard_blocks                           16384
zvol_threads                                      32
```

```
 cat /proc/spl/kstat/zfs/arcstats
5 1 0x01 85 4080 19813013564 313563166597951
name                            type data
hits                            4    279964956
misses                          4    34021474
demand_data_hits                4    35800248
demand_data_misses              4    3984760
demand_metadata_hits            4    174160238
demand_metadata_misses          4    15690254
prefetch_data_hits              4    239568
prefetch_data_misses            4    1619715
prefetch_metadata_hits          4    69764902
prefetch_metadata_misses        4    12726745
mru_hits                        4    47395560
mru_ghost_hits                  4    12884108
mfu_hits                        4    162564934
mfu_ghost_hits                  4    6742441
deleted                         4    14439915
recycle_miss                    4    23210228
mutex_miss                      4    88542
evict_skip                      4    4787839521
evict_l2_cached                 4    0
evict_l2_eligible               4    642752737792
evict_l2_ineligible             4    124734439424
hash_elements                   4    1679310
hash_elements_max               4    1819211
hash_collisions                 4    25552504
hash_chains                     4    434671
hash_chain_max                  4    15
p                               4    0
c                               4    15777049792
c_min                           4    4194304
c_max                           4    10737418240
size                            4    16907171112
hdr_size                        4    540654824
data_size                       4    8776704
meta_size                       4    5604329472
other_size                      4    10753410112
anon_size                       4    21656576
anon_evict_data                 4    0
anon_evict_metadata             4    0
mru_size                        4    4844304384
mru_evict_data                  4    262144
mru_evict_metadata              4    0
mru_ghost_size                  4    10930872832
mru_ghost_evict_data            4    18256384
mru_ghost_evict_metadata        4    10912616448
mfu_size                        4    747145216
mfu_evict_data                  4    262144
mfu_evict_metadata              4    32768
mfu_ghost_size                  4    4827574272
mfu_ghost_evict_data            4    4165240832
mfu_ghost_evict_metadata        4    662333440
l2_hits                         4    0
l2_misses                       4    0
l2_feeds                        4    0
l2_rw_clash                     4    0
l2_read_bytes                   4    0
l2_write_bytes                  4    0
l2_writes_sent                  4    0
l2_writes_done                  4    0
l2_writes_error                 4    0
l2_writes_hdr_miss              4    0
l2_evict_lock_retry             4    0
l2_evict_reading                4    0
l2_free_on_write                4    0
l2_abort_lowmem                 4    0
l2_cksum_bad                    4    0
l2_io_error                     4    0
l2_size                         4    0
l2_asize                        4    0
l2_hdr_size                     4    0
l2_compress_successes           4    0
l2_compress_zeros               4    0
l2_compress_failures            4    0
memory_throttle_count           4    0
duplicate_buffers               4    0
duplicate_buffers_size          4    0
duplicate_reads                 4    0
memory_direct_count             4    17155
memory_indirect_count           4    318295
arc_no_grow                     4    0
arc_tempreserve                 4    0
arc_loaned_bytes                4    0
arc_prune                       4    4352686
arc_meta_used                   4    16898394408
arc_meta_limit                  4    12651942912
arc_meta_max                    4    16912870824
```

```
 cat /proc/spl/kmem/slab
--------------------- cache -------------------------------------------------------  ----- slab ------  ---- object -----  --- emergency ---
name                                    flags      size     alloc slabsize  objsize  total alloc   max  total alloc   max  dlock alloc   max
spl_vn_cache                          0x00020         0         0     4096      104      0     0     1      0     0    21      0     0     0
spl_vn_file_cache                     0x00020         0         0     4096      112      0     0     0      0     0     0      0     0     0
spl_zlib_workspace_cache              0x00240         0         0  8388608   268104      0     0     0      0     0     0      0     0     0
ddt_cache                             0x00040    524288    248560   524288    24856      1     1     2     21    10    42      0     0     0
zio_vdev_cache                        0x00040  67108864  59645040  4194304   131088     16    16    31    480   455   924      0     0     0
zio_buf_20480                         0x00040    524288    348160   524288    20480      1     1    23     21    17   483     46     0     8
zio_data_buf_20480                    0x00040   1048576    716800   524288    20480      2     2    99     42    35  2079      2     0     5
zio_buf_24576                         0x00040    524288    294912   524288    24576      1     1    33     18    12   594     33     0     9
zio_data_buf_24576                    0x00040    524288    442368   524288    24576      1     1    81     18    18  1458      4     0     4
zio_buf_28672                         0x00040         0         0  1048576    28672      0     0    18      0     0   558     33     0     8
zio_data_buf_28672                    0x00040   1048576    802816  1048576    28672      1     1    38     31    28  1178      3     0     1
zio_buf_32768                         0x00040   1048576    917504  1048576    32768      1     1    16     28    28   448     38     0    12
zio_data_buf_32768                    0x00040   1048576    917504  1048576    32768      1     1    29     28    28   812      8     0     2
zio_buf_36864                         0x00040   2097152    958464  1048576    36864      2     2    17     50    26   425     47     0    10
zio_data_buf_36864                    0x00040         0         0  1048576    36864      0     0    28      0     0   700      2     0     1
zio_buf_40960                         0x00040   1048576    778240  1048576    40960      1     1    17     23    19   391     23     0    19
zio_data_buf_40960                    0x00040   1048576     40960  1048576    40960      1     1    25     23     1   575      5     0     3
zio_buf_45056                         0x00040   2097152   1036288  1048576    45056      2     2    18     42    23   378     45     0     9
zio_data_buf_45056                    0x00040         0         0  1048576    45056      0     0    24      0     0   504     15     0     2
zio_buf_49152                         0x00040   1048576     49152  1048576    49152      1     1    29     19     1   551     48     0    42
zio_data_buf_49152                    0x00040   3145728   2801664  1048576    49152      3     3    23     57    57   437      8     0     1
zio_buf_53248                         0x00040   1048576    798720  1048576    53248      1     1    27     18    15   486     38     0    10
zio_data_buf_53248                    0x00040         0         0  1048576    53248      0     0    28      0     0   504      6     0     2
zio_buf_57344                         0x00040   1048576    802816  1048576    57344      1     1    25     17    14   425     46     0    19
zio_data_buf_57344                    0x00040         0         0  1048576    57344      0     0    30      0     0   510      7     0     3
zio_buf_61440                         0x00040   6291456   4239360  2097152    61440      3     3    12     93    69   372     37     0     8
zio_data_buf_61440                    0x00040         0         0  2097152    61440      0     0    13      0     0   403      8     0     1
zio_buf_65536                         0x00040   2097152   1966080  2097152    65536      1     1    12     30    30   360     26     0    16
zio_data_buf_65536                    0x00040   4194304    131072  2097152    65536      2     2    17     60     2   510     19     0     3
zio_buf_69632                         0x00040   2097152   1810432  2097152    69632      1     1    13     28    26   364     44     0     4
zio_data_buf_69632                    0x00040         0         0  2097152    69632      0     0    17      0     0   476     22     0     5
zio_buf_73728                         0x00040         0         0  2097152    73728      0     0    15      0     0   390     31     0    11
zio_data_buf_73728                    0x00040         0         0  2097152    73728      0     0    15      0     0   390      3     0     1
zio_buf_77824                         0x00040   6291456   5836800  2097152    77824      3     3    15     75    75   375     37     0    16
zio_data_buf_77824                    0x00040         0         0  2097152    77824      0     0    13      0     0   325     10     0     4
zio_buf_81920                         0x00040         0         0  2097152    81920      0     0    18      0     0   432     18     0    26
zio_data_buf_81920                    0x10040         0         0  2097152    81920      0     0    15      0     0   360      5     0     2
zio_buf_86016                         0x00040   2097152   1978368  2097152    86016      1     1    16     23    23   368     15     0     4
zio_data_buf_86016                    0x00040         0         0  2097152    86016      0     0    16      0     0   368      7     0     2
zio_buf_90112                         0x00040         0         0  2097152    90112      0     0    17      0     0   374     31     0     4
zio_data_buf_90112                    0x10040         0         0  2097152    90112      0     0    20      0     0   440      2     0     1
zio_buf_94208                         0x10040         0         0  2097152    94208      0     0    21      0     0   441     14     0    44
zio_data_buf_94208                    0x00040         0         0  2097152    94208      0     0    15      0     0   315      3     0     1
zio_buf_98304                         0x00040         0         0  2097152    98304      0     0    22      0     0   440     11     0    10
zio_data_buf_98304                    0x10040         0         0  2097152    98304      0     0    17      0     0   340      2     0     1
zio_buf_102400                        0x00040         0         0  2097152   102400      0     0    18      0     0   342      8     0     8
zio_data_buf_102400                   0x10040   2097152    102400  2097152   102400      1     1    16     19     1   304      6     0     1
zio_buf_106496                        0x00040   2097152    106496  2097152   106496      1     1    20     18     1   360     20     0     5
zio_data_buf_106496                   0x00040   2097152   1916928  2097152   106496      1     1    17     18    18   306     12     0     7
zio_buf_110592                        0x00040         0         0  2097152   110592      0     0    19      0     0   342     15     0     5
zio_data_buf_110592                   0x00040   2097152   1990656  2097152   110592      1     1    18     18    18   324      2     0     1
zio_buf_114688                        0x00040         0         0  2097152   114688      0     0    22      0     0   374     15     0     6
zio_data_buf_114688                   0x00040   2097152   1949696  2097152   114688      1     1    21     17    17   357      2     0     1
zio_buf_118784                        0x10040         0         0  2097152   118784      0     0    11      0     0   187      2     0     1
zio_data_buf_118784                   0x10040   2097152   2019328  2097152   118784      1     1    16     17    17   272      2     0     1
zio_buf_122880                        0x00040         0         0  2097152   122880      0     0    11      0     0   176      0     0     0
zio_data_buf_122880                   0x00040   2097152   1843200  2097152   122880      1     1    22     16    15   352      2     0     1
zio_buf_126976                        0x00040         0         0  4194304   126976      0     0     8      0     0   248      5     0     1
zio_data_buf_126976                   0x00040         0         0  4194304   126976      0     0    11      0     0   341      7     0     1
zio_buf_131072                        0x00040  25165824  17170432  4194304   131072      6     6    37    186   131  1132      0     0     0
zio_data_buf_131072                   0x00040  20971520  18874368  4194304   131072      5     5  1037    155   144 32147     10     0     1
spill_cache                           0x00040         0         0  4194304   131072      0     0     0      0     0     0      0     0     0
zfs_znode_cache                       0x00020 7130841088 6714732696    32768     1064  217616 217616 217753  6310864 6310839 6314837      0     0     0
```

**Comment by dweeezil** [comment]
@asbach It doesn't look like 0.6.3-5~trusty has 3c832b8 which I think might help this.

**Comment by asbach** [comment]
Any idea when this will show up in the trusty builds?

**Comment by behlendorf** [comment]
@asbach one solution would be to roll back your kernel.  An interface change on the kernel side is the root cause here.  When that fix might appear for trusty is up to @dajhorn.

**Comment by dajhorn** [comment]
The dailies will begin updating with the next Ubuntu series, which is scheduled for February 27th.  Anything added to the stable branch should be reflected in the stable PPA within two weeks.

**Comment by asbach** [comment]
I experienced the problem using the 3.13.0-44 and 3.13.0-45 kernels. I've reverted to the 3.13.0-43 and so far it has been working quite well. I can say for sure, once the work week starts again and the server has its usual load. 
In my quest to get this problem in check I've set my arc_max to 10GB, disabled atime and also set the spl_kmem_cache_slab_limit to 16384. I've also set the vm.min_free_kbytese to 1048576, since I thought arc and kswap were fighting over memory. The changes didn't havy any effect on the original problem.
